### PR TITLE
Improves status handling for jupyter-controller

### PR DESCRIPTION
**Pull request by DnPlas**
Following the standard for status handling, this PR introduces a couple of changes in the charm code to improve how unit status are managed.

### CODE REVIEW

**Comment by DomFleischmann** [comment]
Code looks LGTM, but please fix the linting errors.

File: charms/jupyter-controller/tests/test_operator.py [hunk 1]
```diff
@@ -11,7 +11,7 @@ def harness():
 
 def test_not_leader(harness):
     harness.begin()
-    assert harness.charm.model.unit.status == WaitingStatus("Waiting for leadership")
+    assert harness.charm.model.unit.status == MaintenanceStatus("")
```
Comment (DomFleischmann) [inline charms/jupyter-controller/tests/test_operator.py:14]: Given that this status check is now done in `main` you will have to call `harness.begin_with_inital_hooks` else you never reach that logic.

File: charms/jupyter-controller/src/charm.py [hunk 2]
```diff
@@ -13,31 +13,42 @@
...
 
 
+class CheckFailed(Exception):
+    """ Raise this exception if one of the checks in main fails. """
+
+    def __init__(self, msg, status=None):
+        super().__init__()
+
+        self.msg = msg
+        self.status = status
```
Comment (DomFleischmann) [inline charms/jupyter-controller/src/charm.py:23]: Can you update this with the latest changes we added? You can see them [here](https://github.com/canonical/argo-operators/blob/2f7d7d4ce9cd3031e115a458a624f78cdd4d5aca/operators/argo-controller/src/charm.py#L23) and [here](https://github.com/canonical/argo-operators/blob/2f7d7d4ce9cd3031e115a458a624f78cdd4d5aca/operators/argo-controller/src/charm.py#L61)

**Review by DomFleischmann — CHANGES_REQUESTED** [review]

File: charms/jupyter-controller/tests/test_operator.py (hunk 1 above)
Comment (DnPlas) [reply charms/jupyter-controller/tests/test_operator.py]: Updated.

File: charms/jupyter-controller/src/charm.py (hunk 2 above)
Comment (DnPlas) [reply charms/jupyter-controller/src/charm.py]: I believe it looks like the latest version of your PRs, let me know otherwise.

**Review by DomFleischmann — APPROVED** [review]

**Pull request merged**